### PR TITLE
Use JSON instead of PSTORE for module metadata

### PR DESCRIPTION
**Pull request by clee-r7**
Convert metadata pstore(binary) to json(text) to reduce repository history footprint and thus size.

## Verification

- [ ] Start `msfconsole`
- [ ] **Verify** that ~/.msf4/store/modules_metadata.json is generated
- [ ] **Verify** that a search for your favorite module returns correctly - **EG: `search bitcoin`** should return 1 result.

### CODE REVIEW

File: lib/msf/core/modules/metadata/cache.rb [hunk 1]
```diff
@@ -35,10 +35,10 @@ def refresh_metadata_instance(module_instance)
   #  Returns the module data cache, but first ensures all the metadata is loaded
   #
   def get_metadata
-    @mutex.synchronize {
+   @mutex.synchronize {
```
Comment (mkienow-r7) [inline lib/msf/core/modules/metadata/cache.rb]: Minor: Why was indent reduced by one here and few lines below?

File: lib/msf/core/modules/metadata/obj.rb [hunk 2]
```diff
@@ -26,7 +27,12 @@ class Obj
   attr_reader :is_install_path
   attr_reader :ref_name
 
-  def initialize(module_instance)
+  def initialize(module_instance, obj_hash = nil)
+    unless (obj_hash.nil?)
```
Comment (mkienow-r7) [inline lib/msf/core/modules/metadata/obj.rb]: Minor: style guide - no need for parentheses here

File: lib/msf/core/modules/metadata/obj.rb [hunk 3]
```diff
@@ -65,6 +111,44 @@ def path
...
+    @mod_time = Time.parse(obj_hash['mod_time'])
+    @ref_name = obj_hash['ref_name']
+    @path = obj_hash['path']
+    @is_install_path = obj_hash['is_install_path']
+    @targets = obj_hash['targets'].nil? ? [] : obj_hash['targets']
+  end
+
+  def sort_platform_string
+    arr = @platform.split(',')
+    if !arr.nil? && arr.length > 1
```
Comment (bcoles) [inline lib/msf/core/modules/metadata/obj.rb]: `arr` is initialised with a call to `split` on the previous line. A call to `split` should never return `nil`, so `arr` should never be `nil`, making the `nil` check redundant.

If the `split` call fails, you have bigger problems.

Also, `size` is preferred over `length` for array length comparison, but it's a [suggestion](https://github.com/rubocop-hq/ruby-style-guide#map-find-select-reduce-size), not a requirement.


```ruby
  def sort_platform_string
    arr = @platform.split(',')
    @platform = arr.sort.join(',') if arr.size > 1
  end
```

Comment (clee-r7) [reply lib/msf/core/modules/metadata/obj.rb]: thanks

File: lib/msf/core/modules/metadata/cache.rb (hunk 1 above)
Comment (clee-r7) [reply lib/msf/core/modules/metadata/cache.rb]: its was off previously

File: lib/msf/core/modules/metadata/store.rb [hunk 4]
```diff
@@ -28,10 +32,13 @@ def init_store
...
-        @store[:module_metadata] = @module_metadata_cache
-      end
-    rescue Excepion => e
+      @update_mutex.synchronize {
+        json_map = @module_metadata_cache.sort.to_h
+        File.open(@path_to_user_metadata, "w") do |f|
+          f.write(JSON.pretty_generate(json_map))
+        end
+      }
+    rescue Exception => e
```
[the remark this answers is not in the document]
Comment (clee-r7) [reply lib/msf/core/modules/metadata/store.rb]: I guess not

**Comment by jmartin-r7** [comment]
Unfortunately testing shows this still persists the original problem, this `json` file will change on first install from using the `git` repo.

```
106113,106116c102311,102312
<     "targets": [
<
<     ],
<     "mod_time": "2017-08-04 22:13:59 +0000",
---
>     "targets": null,
>     "mod_time": "2017-08-23 11:34:40 -0500",
106139,106142c102335,102336
<     "targets": [
<
<     ],
<     "mod_time": "2018-05-31 20:45:51 +0000",
---
>     "targets": null,
>     "mod_time": "2018-05-15 10:41:17 -0500",
106165,106168c102359,102360
<     "targets": [
<
<     ],
<     "mod_time": "2018-05-31 20:45:51 +0000",
---
>     "targets": null,
>     "mod_time": "2017-11-27 11:58:44 -0600",
```

Reproduced with:
```
rm -rf ~/.msf4

cat > waitForStore.rc << EOF
<ruby>
require "msf/core/modules/metadata/store"

UserMetaDataFile = Msf::Modules::Metadata::Store::UserMetaDataFile
store_dir = File.join(Msf::Config.config_directory, "store")
cache_file = File.join(store_dir, UserMetaDataFile)
while true do
break if File.exist?(cache_file)
sleep(5)
end
time = File.stat(cache_file).mtime
count = 0
framework.modules.refresh_cache_from_module_files
while true do
  new_time = File.stat(cache_file).mtime
  count += 1
  break unless new_time > time
  break unless count < 30
  time = new_time
  sleep(5)
end
</ruby>
exit
EOF

./msfconsole -qr waitForStore.rc
```

**Comment by jmartin-r7** [comment]
It has been determined that the `json` serializer changed to produce the `target` changes.

File modification times are determined by the specific installation by adding to the update task we can avoid these changes in the packaging update job.
```
git ls-files modules/ -z | xargs -0 -n1 -I{} -- git log -1 --format="%ai {}" {} | while read -r udate utime utz ufile ; do
  touch -d "$udate $utime" $ufile
done
``` 

A new commit will be added with a `json` file serialized for the `targets` issue.

**Comment by jmartin-r7** [comment]
## Release Notes

The revised module cache implementation has migrated from a pstore(binary) file to a JSON format to reduce repository history footprint and thus size.

**Pull request merged**